### ISSUE TITLE
Very low performance in SQL store when comparing with file store

### ISSUE THREAD

**Issue by alanyeung95** [nats-io/nats-streaming-server#899]
I have tried the [benchmark example](https://github.com/nats-io/stan.go/tree/master/examples/stan-bench) to see the comparison between different persistent store. And here is the (average) results on my mac machine.

For the sql store, I just use the storage & service on my host rather than in docker container. I didn't config other setting on mysq/postgesql. Here are the results:

## FILE store
```
# nats-streaming-server -c filestore.conf
```
```
Starting benchmark [msgs=100000, msgsize=1024, pubs=1, subs=1]
NATS Streaming Pub/Sub stats: 34,601 msgs/sec ~ 33.79 MB/sec
 Pub stats: 65,580 msgs/sec ~ 64.04 MB/sec
 Sub stats: 17,300 msgs/sec ~ 16.90 MB/sec
```
## POSTGESQL store with caching
```
# nats-streaming-server --store=SQL --sql_driver=postgres --sql_no_caching=false --sql_source='user=user host=localhost port=5432 dbname=nss_db sslmode=disable'
```
```
Starting benchmark [msgs=100000, msgsize=1024, pubs=1, subs=1]
NATS Streaming Pub/Sub stats: 3,350 msgs/sec ~ 3.27 MB/sec
 Pub stats: 2,609 msgs/sec ~ 2.55 MB/sec
 Sub stats: 1,675 msgs/sec ~ 1.64 MB/sec
```
## MYSQL store with caching
```
nats-streaming-server --store=SQL --sql_driver=mysql --sql_no_caching=false --sql_source='nss:password@(127.0.0.1:3306)/nss_db' 
sql_no_caching
```
```
Starting benchmark [msgs=100000, msgsize=1024, pubs=1, subs=1]
NATS Streaming Pub/Sub stats: 3,180 msgs/sec ~ 3.11 MB/sec
 Pub stats: 3,325 msgs/sec ~ 3.25 MB/sec
 Sub stats: 1,590 msgs/sec ~ 1.55 MB/sec
```

So basically the filestore is around 10 times faster than the sqlstore. I wonder **if the NATS development team share the similar result**. If not, is there any **config with the sql store** we can still use apart from `sql_no_caching`? 

10 times slower is really a blocker from using sqlstore when massive messages need to be transmitted even RDBMS has its advantages.

**Comment by kozlovic** [comment]
Yes, there is a performance difference between our File and SQL implementations. When messages come from the same connection and using a sync publish call, this will result in an INSERT to the message table, and if a subscriber is running, that will result in an INSERT to the subscription's pending list table too. When you use async publish, or have multiple connections, it is possible to "batch" some inserts and flush them. Still, performance is not at the level of the FileStore for sure.
I would welcome DB experts to help tune DBs for Streaming or review our code and suggest improvements. But as it stands, your observation is correct.

**Comment by alanyeung95** [comment]
Thank you for the information, I think FileStore is more suitable for my project right now.
